File: libraries/listing/src/main/kotlin/orwir/starrit/listing/feed/Post.kt
package orwir.starrit.listing.feed

import android.content.res.Resources
import android.net.Uri
import androidx.core.net.toUri
import orwir.starrit.core.extension.toHtml
import orwir.starrit.core.util.createHashCode
import orwir.starrit.core.util.prettyDate
import orwir.starrit.core.util.squeeze
import orwir.starrit.listing.R
import orwir.starrit.listing.model.Submission
import orwir.starrit.listing.model.Subreddit
import orwir.starrit.listing.util.imageUrlOrNull
import orwir.starrit.listing.util.isImageUrl
import orwir.starrit.listing.util.videoOrNull

sealed class Post(submission: Submission, res: Resources) {
    val id: String = submission.id
    val subreddit: Subreddit = submission.subreddit
    val author: String = submission.author
    val created: String = prettyDate(submission.created * 1000, res)
    val title: String = submission.title
    val nsfw: Boolean = submission.nsfw
    val spoiler: Boolean = submission.spoiler
    val comments: String = submission.commentsCount.squeeze()
    val score: String = prettyScore(submission, res)
    val domain: String = submission.domain
    val selfDomain: Boolean = domain.startsWith("self.")
    val contentUrl = submission.url
    val postUrl = submission.permalink

    val imagePreview: String = submission.preview
        ?.images
        ?.firstOrNull()
        ?.resolutions
        ?.firstOrNull()
        ?.url
        ?: submission.thumbnail.takeIf(::isImageUrl)
        ?: ""

    val imageSource: String = submission.preview
        ?.images
        ?.firstOrNull()
        ?.source
        ?.url
        ?: submission.imageUrlOrNull()
        ?: ""

    override fun equals(other: Any?) =
        other is Post
                && other::class == this::class
                && id == other.id
                && subreddit == other.subreddit
                && author == other.author
                && created == other.created
                && title == other.title
                && nsfw == other.nsfw
                && spoiler == other.spoiler
                && comments == other.comments
                && score == other.score
                && domain == other.domain
                && contentUrl == other.contentUrl
                && postUrl == other.postUrl

    override fun hashCode() = createHashCode(
        id,
        subreddit,
        author,
        created,
        title,
        nsfw,
        spoiler,
        comments,
        score,
        domain,
        contentUrl,
        postUrl
    )

    private fun prettyScore(submission: Submission, res: Resources) =
        if (submission.hideScore) {
            res.getString(R.string.score_hidden)
        } else {
            submission.score.squeeze()
        }
}

class LinkPost(
    submission: Submission,
    resources: Resources
) : Post(submission, resources) {
    val link: Uri = contentUrl.toUri()
    val displayLink: String = link.authority ?: submission.url
}

class ImagePost(
    submission: Submission,
    resources: Resources
) : Post(submission, resources) {
    private val image = submission.preview
        ?.images
        ?.firstOrNull()
        ?.resolutions
        ?.firstOrNull()
    val imageWidth: Int = image?.width ?: 0
    val imageHeight: Int = image?.height ?: 0
}

class GifPost(
    submission: Submission,
    resources: Resources
) : Post(submission, resources) {
    val gif: String = submission.url
}

class TextPost(
    submission: Submission,
    resources: Resources
) : Post(submission, resources) {
    val text: CharSequence = (submission.selftextHtml ?: submission.selftext ?: "").toHtml()
}

class VideoPost(
    submission: Submission,
    resources: Resources
) : Post(submission, resources) {
    val video: String = submission.videoOrNull() ?: ""
}

File: lib/models/post.dart
import 'package:flutter/foundation.dart';
import 'package:starrit/utils/json.dart';
import 'package:starrit/utils/object.dart';
import 'subreddit.dart';
import 'author.dart';
import 'image.dart';

enum PostType { image, gif, video, text, link }

@immutable
class Post {
  Post({
    @required this.id,
    @required this.subreddit,
    @required this.author,
    @required this.created,
    @required this.title,
    @required this.nsfw,
    @required this.spoiler,
    @required this.comments,
    @required this.score,
    @required this.domain,
    @required this.selfDomain,
    @required this.postUrl,
    @required this.contentUrl,
    @required this.type,
    @required this.imagePreview,
    @required this.imageSource,
    @required this.imageBlurred,
    @required this.text,
    @required this.gif,
    @required this.video,
  });

  Post.fromJson(Map<String, dynamic> json)
      : this(
          id: json['name'],
          subreddit: Subreddit.fromJson(json),
          author: Author.fromJson(json),
          created: json.created,
          title: json['title'],
          nsfw: json['over_18'] ?? false,
          spoiler: json['spoiler'] ?? false,
          comments: json['num_comments'],
          score: json['score'],
          domain: json.domain,
          selfDomain: json.selfDomain,
          postUrl: json['permalink'],
          contentUrl: json.url,
          type: json.type,
          imagePreview: json.image('preview.images[0].resolutions[0]'),
          imageSource: json.image('preview.images[0].source'),
          imageBlurred: json.image(
            'preview.images[0].variants.nsfw.resolutions[0]',
          ),
          text: json.text,
          gif: json.gif,
          video: json.video,
        );

  final String id;
  final Subreddit subreddit;
  final Author author;
  final DateTime created;
  final String title;
  final bool nsfw;
  final bool spoiler;
  final int comments;
  final int score;
  final String domain;
  final bool selfDomain;
  final String postUrl;
  final String contentUrl;
  final PostType type;
  final PostImage imagePreview;
  final PostImage imageSource;
  final PostImage imageBlurred;
  final String text;
  final String gif;
  final String video;

  @override
  int get hashCode => hash([
        id,
        subreddit,
        author,
        created,
        title,
        nsfw,
        spoiler,
        comments,
        domain,
        postUrl,
        contentUrl,
        type,
        imagePreview,
        imageSource,
        imageBlurred,
        text,
        gif,
        video,
      ]);

  @override
  bool operator ==(Object other) {
    return identical(this, other) ||
        other is Post &&
            runtimeType == other.runtimeType &&
            id == other.id &&
            subreddit == other.subreddit &&
            author == other.author &&
            created == other.created &&
            title == other.title &&
            nsfw == other.nsfw &&
            spoiler == other.spoiler &&
            comments == other.comments &&
            domain == other.domain &&
            postUrl == other.postUrl &&
            contentUrl == other.contentUrl &&
            type == other.type &&
            imagePreview == other.imagePreview &&
            imageSource == other.imageSource &&
            imageBlurred == other.imageBlurred &&
            text == other.text &&
            gif == other.gif &&
            video == other.video;
  }
}

extension _ on Map<String, dynamic> {
  DateTime get created => DateTime.fromMillisecondsSinceEpoch(
        get<double>('created_utc').toInt() * 1000,
        isUtc: true,
      );

  String get domain => string('domain');

  bool get selfDomain => domain.startsWith('self.');

  String get url => string('url');

  PostType get type {
    if (isImage) return PostType.image;
    if (video != null) return PostType.video;
    if (gif != null) return PostType.gif;
    if (text != null || selfDomain) return PostType.text;
    return PostType.link;
  }

  PostImage image(String key) {
    return get<Object>(key)?.into((json) => PostImage.fromJson(json));
  }

  String get text => string('selftext') ?? string('selftext_html');

  String get gif => url.endsWith('.gif') ? url : null;

  String get video {
    if (domain == 'i.imgur.com' && url.endsWith('.gifv')) {
      return url.replaceAll('http://', 'https://').replaceAll('.gifv', '.mp4');
    }
    if (domain == 'v.redd.it') {
      final key = 'secure_media.reddit_video.hls_url';
      return string(key) ?? string('crosspost_parent_list[0].$key');
    }
    if (domain == 'gfycat.com') {
      final key = 'secure_media.oembed.thumbnail_url';
      final url = string(key) ?? string('crosspost_parent_list[0].$key');
      if (url != null) {
        return url
            .replaceAll('thumbs.gfycat.com', 'giant.gfycat.com')
            .replaceAll('-size_restricted', 'replace')
            .replaceAll('.gif', '.mp4');
      }
    }
    return null;
  }

  bool get isImage {
    final lastSegment = Uri.parse(string('url')).pathSegments.last;
    final ext = lastSegment?.substring(lastSegment.lastIndexOf('.') + 1);
    return _imageExtensions.contains(ext?.toLowerCase());
  }
}

const _imageExtensions = [
  'bmp',
  'jpg',
  'jpeg',
  'png',
  'svg',
  'tif',
  'tiff',
  'jfif',
  'pjpeg',
  'pjp',
  'ico',
  'cur',
];
